src/mlpack/methods/ann/layer/grouped_convolution_impl.hpp

```diff
@@ -0,0 +1,690 @@
+      {
+        // Iterate over input maps (we will apply the filter and sum).
+        for (size_t inMap = 0; inMap < inGroupSize; ++inMap)
+        {
+          MatType output;
+          GradientConvolutionRule::Convolution(
+              inputTemp.slice((group * inGroupSize) + inMap + fullInputOffset),
+              mappedError.slice(outMap + fullOutputOffset),
+              output,
+              strideWidth,
+              strideHeight);
+
+          // TODO: understand this conditional.  Is it needed?
+          if (gradientTemp.n_rows < output.n_rows ||
+              gradientTemp.n_cols < output.n_cols)
+          {
+            gradientTemp.slice((outMap * inGroupSize) + inMap) += 
+                output.submat(
+                    0,
+                    0,
+                    gradientTemp.n_rows - 1, 
+                    gradientTemp.n_cols - 1);
+          }
+          else if (gradientTemp.n_rows > output.n_rows ||
+                  gradientTemp.n_cols > output.n_cols)
+          {
+            gradientTemp.slice((outMap * inGroupSize) + inMap).submat(0, 0,
+                output.n_rows - 1, output.n_cols - 1) += output;
+          }
+          else
+          {
+            gradientTemp.slice((outMap * inGroupSize) + inMap) += output;
+          }
+        }
+
+        if (useBias)
+          gradient[weight.n_elem + outMap] += arma::accu(mappedError.slice(
+              outMap + fullOutputOffset));
+      }
+    }
+  }
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename MatType
+>
+void GroupedConvolutionType<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    MatType
+>::ComputeOutputDimensions()
+{
+  // First, we must make sure the padding sizes are up to date, which we can
+  // now do since inputDimensions is set correctly.
+  if (paddingType == "valid")
+  {
+    padWLeft = 0;
+    padWRight = 0;
+    padHTop = 0;
+    padHBottom = 0;
+  }
+  else if (paddingType == "same")
+  {
+    InitializeSamePadding();
+  }
+
+  padding = ann::Padding(padWLeft, padWRight, padHTop, padHBottom);
+  padding.InputDimensions() = this->inputDimensions;
+  padding.ComputeOutputDimensions();
+
+  // We must ensure that the output has at least 3 dimensions, since we will
+  // be adding some number of maps to the output.
+  this->outputDimensions = std::vector<size_t>(
+      std::max(this->inputDimensions.size(), size_t(3)), 1);
+  this->outputDimensions[0] = ConvOutSize(this->inputDimensions[0],
+      kernelWidth, strideWidth, padWLeft, padWRight);
+  this->outputDimensions[1] = ConvOutSize(this->inputDimensions[1],
+      kernelHeight, strideHeight, padHTop, padHBottom);
+
+  inMaps = (this->inputDimensions.size() >= 3) ? this->inputDimensions[2] : 1;
+
+  if (groups == 0)
+    Log::Fatal << "Groups must be greater than 0." << std::endl;
```

Comment:
```suggestion
    Log::Fatal << "GroupedConvolution::ComputeOutputDimensions(): groups must "
        << "be greater than 0." << std::endl;
```
It's probably good to prepend the name of the class to the error message, since a user may not actually get a backtrace of any sort.